[Clang][IR] add TBAA metadata on pointer, union and array types.

dybv-sc opened the pull request:
Options to disable new behaviour:

-Xclang -no-union-tbaa
-Xclang -no-pointer-tbaa
-Xclang -no-array-tbaa

To enable union struct path tbaa there was need to update the way llvm handles struct path tbaa metadata. Multiple fields with same offset are now allowed. To properly resolve access type, struct path tbaa visitor considers each field with given offset. This patch works both for new and old struct path TBAA.

Also this patch enables tbaa aware array and pointer accesses by default.

Endilll commented on clang/test/CXX/drs/dr158.cpp:
```diff
@@ -1,7 +1,7 @@
-// RUN: %clang_cc1 -triple x86_64-linux -std=c++98 %s -O3 -disable-llvm-passes -pedantic-errors -emit-llvm -o - | FileCheck %s
```
Can you explain why `-disable-llvm-passes` is there, and why is can be removed now?

dybv-sc replied:
Yes,

The matter is that if no llvm passes are run, IR is unoptimized and contains extra load and store instruction that operate with stack (storing and loading params there). Those extra loads should not be considered by the test, because it's goal to check tbaa for specific load and store coming from line 22 and 24. Before I introduced more precise metadata on pointer types, all of that load and stores had same metadata, so test passed even if wrong load was considered. Now, this test should pick only load coming from line 22. So, to make things easier, I just enabled optimization passes and final IR contains only one load and one store instruction.

rjmccall replied:
This is okay as long as you're sure the optimization will be performed and you make sure your test only checks the pattern you want out of the optimizer and doesn't cover other details that could potentially be changed by unrelated LLVM patches.  Otherwise, we generally aim to make IR-generation tests "unit" tests that specifically test the output of Clang's IR-generation.  For output that's only included in optimized modes, like TBAA metadata, that generally requires enabling optimization (with `-O<n>`) and then disabling all the LLVM optimizations (with `-disable-llvm-passes`).

dybv-sc replied:
I am sure that level 3 optimization should cut out all stack manipulation here leaving only one necessary load(const int *x = *p) and one necessary store(*q = 0). This test checks only presence of that load/store and metadata information, so I don't think it could be disrupted by unrelated patches.

rjmccall commented on clang/lib/CodeGen/CodeGenTBAA.cpp:
```diff
@@ -184,13 +199,24 @@ llvm::MDNode *CodeGenTBAA::getTypeInfoHelper(const Type *Ty) {
-  // pointers distinct.
-  if (Ty->isPointerType() || Ty->isReferenceType())
-    return createScalarTypeNode("any pointer", getChar(), Size);
+  // Pointer types never alias if their pointee type is distinct.
```
This comment both is incorrect and doesn't match the code.  What you seem to be implementing here is the C++ similar-type rule, in which what matters is whether the pointer type sub-structure matches while ignoring qualifiers.

Unfortunately, this rule seems to treat `void*` and `char*` as different types.  That is wrong in C because those types are compatible, and we probably ought to use the C rule even in C++.

Also, in general, I would suggest that you write `getPointeeName` much more conservatively rather than assuming that you can just render an arbitrary type and it's going to be okay to treat different renderings as distinct for TBAA purposes.  For example, this is going to treat pointers to different vector types as non-aliasing, which worries me a lot because vector programmers are often pretty fast-and-loose.  You're also going to be stricter about ObjC than I'm comfortable with.  I would strongly suggest just doing this rendering for specific kinds of types, like records, and otherwise having some kind of fallback to the any-pointer metadata.

dybv-sc replied:
I agree that this may be a little bit too strict, but, as far as I could understand, according to C11 standard `void*` and `char*` are not the same (correct me if I am wrong):

From section 6.5 point 7:
> an object shall have its stored value accessed only by an lvalue expression that has one of
> the following types:88)
> — a type compatible with the effective type of the object,
> — a qualified version of a type compatible with the effective type of the object,
> — a type that is the signed or unsigned type corresponding to the effective type of the
> object,
> — a type that is the signed or unsigned type corresponding to a qualified version of the
> effective type of the object,
> — an aggregate or union type that includes one of the aforementioned types among its
> members (including, recursively, a member of a subaggregate or contained union), or
> — a character type.

From all of above, the only way `void*` and `char*` alias if they are compatible, but they are not:

From section 6.2.7:
> Two types have compatible type if their types are the same. Additional rules for
> determining whether two types are compatible are described in 6.7.2 for type specifiers,
> in 6.7.3 for type qualifiers, and in 6.7.6 for declarators.

And from section 6.7.6.1 for pointer declarators:

> For two pointer types to be compatible, both shall be identically qualified and both shall
> be pointers to compatible types.

`void` and `char` types are not compatible, because they are not the same. That means that `void*` and `char*` are not compatible and thus shall not alias.

So, from what I see in standard, pointer types to incompatible types should not alias. May be I got something wrong, so let's discuss it.

Also, now after I revisited standard, I see that getPointeeName is really not so accurate - it should consider type compatibility better(not only by exact match).

rjmccall commented on clang/lib/CodeGen/CodeGenTBAA.cpp:
```diff
@@ -184,13 +199,24 @@ llvm::MDNode *CodeGenTBAA::getTypeInfoHelper(const Type *Ty) {
+    do {
+      PtrDepth++;
+      Ty = Ty->getPointeeType().getTypePtr();
+    } while (!Ty->getPointeeType().isNull());
```
`getPointeeType()` will look through a lot of types that you probably don't want to look through, including member pointers.  You should write this to specifically look for pointers.  (References can't occur in nested positions.)

dybv-sc replied:
Actually I want to look for member pointers. In [conv.qual] of C++ standard mentioned:

> each Pi is “pointer to” (9.3.4.2), “pointer to member of
> class Ci of type” (9.3.4.4), “array of Ni”, or “array of unknown bound of” (9.3.4.5

Where Pi is i-th inderection in qualification decomposition. I want to consider all those types of indirection.
I think, all I do if I get rid of getPointeeType() is just move most of it's checks outside.

rjmccall replied:
You're right, you do need to look through member pointers here, although this means you're also treating member-pointer structure as if it were just pointer structure.  I guess that's okay for now.

rjmccall commented on clang/lib/CodeGen/CodeGenTBAA.cpp:
```diff
@@ -105,13 +105,28 @@ static bool isValidBaseType(QualType QTy) {
+  }
+
+  if (Ty->isIncompleteType())
+    return "<incomplete type>";
```
Why are incomplete types treated differently here?

dybv-sc replied:
Removed. I was uncertain in a moment what to do in that case, but after revisiting C/C++ standard I learned that there is no difference between complete and incomplete types when considering their similarity/compatibility.

rjmccall reviewed (commented):
In general, this patch needs to be clearer about what rules it's actually enforcing.  You're adding new command-line options, but users have to guess what they mean!

If you're going to be working on TBAA, would you mind adding a section to Clang's manual (`UsersManual.rst`) about type-based alias analysis?  We should start by documenting our current behavior, then document the behavior of all these new options.  To make this a less onerous request, let me suggest some starter text:

```
C and C++ forbid programmers from accessing objects using l-values that don't match the type of the object.  By default, Clang takes advantage of these rules to decide that certain pointers cannot point to the same object; this is called *strict aliasing* or *type-based alias analysis* (TBAA).  This can be completely disabled using the option ``-fno-strict-aliasing``.  ``-fno-strict-aliasing`` is the default for ``clang-cl``.

When strict aliasing is enabled, Clang uses the type-based aliasing rules from the appropriate standard for the current language mode.  In the C standard, the aliasing rules are laid out in section 6.5 (Expressions).  In the C++ standard, the aliasing rules are laid out in [basic.lval].  For the most part, the C and C++ rules coincide and can be summarized as follows:

- An object can be accessed through an l-value of character type (e.g. ``char``).
- An object of integer type can be accessed through an l-value of different signedness; e.g. a ``signed short`` object can be accessed through an ``unsigned short`` l-value.
- Otherwise, objects can only be accessed through l-values of the type of the object.

For the exact rules, please consult the standards.  Clang generally reserves the flexibility to take advantage of the exact rules for the current language mode, except as noted here:

- While C gives all character types the power to arbitrarily alias, C++ reserves this to ``char`` and ``unsigned char``.  Clang relaxes this rule in C++ to match the C rule.

There are several ways to load from or store to an object as if it had a different type without violating the strict aliasing rule.  The most explicit and portable is to ``memcpy`` between the object and an object of the desired type; for aliasing purposes, ``memcpy`` behaves as if it used loads and stores of character type.  Clang also supports ``__attribute__((may_alias))``, which can be placed on a type declaration (such as a ``struct`` or ``typedef``) to give that type the equivalent aliasing power of a character type.

Clang uses an implementation model in which "sufficiently obvious" aliasing should override type-based assumptions.  Strict aliasing means that Clang will assume that `int*` and `float*` parameters to a function do not alias, and it may reorder loads and stores to those parameters accordingly.  However, if a `float*` parameter to a function is cast to `int*`, Clang will understand that the result of the cast still aliases the original parameter, and it should not reorder loads and stores to those pointers. This is only a best-effort attempt to avoid miscompiles, and programmers should generally still aim to write code which does not violate the strict aliasing rules, as discussed above.

An access to a member of an aggregate type (such as a ``struct``) is considered to also be an access to the aggregate.  This means that there must also be an object of the aggregate type at that location, and it means that accesses into different aggregates cannot alias.  This rule can be weakened to only consider the final accessed type using ``-fno-struct-path-tbaa``.

<document your new options here>
```

dybv-sc commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -4598,8 +4602,7 @@ LValue CodeGenFunction::EmitLValueForField(LValue base,
     FieldTBAAInfo = TBAAAccessInfo::getMayAliasInfo();
-  } else if (rec->isUnion()) {
-    // TODO: Support TBAA for unions.
+  } else if (rec->isUnion() && !CGM.getCodeGenOpts().UnionTBAA) {
```
The goal of that patch is to re-utilize existing struct path TBAA metadata and amend it functionality to work for union types. Here I allow for that metadata to be generated for union members accesses, so yes there will be tbaa struct records with all members with offset 0. Later in code I adapted current struct-path-tbaa tree walkers to handle such situations where multiple fields have same offset.

rjmccall commented:
> From all of above, the only way void* and char* alias if they are compatible, but they are not:

Ah, you're right, I had misremembered this.  6.2.5p31 requires them to have the same representation as each other, and there's a footnote about this being meant to imply "interchangeability as arguments to functions, return values from functions, and members of unions", but it doesn't go so far as to make them formally compatible.  And even if we wanted to read that as implying a sort of semi-compatibility that ought to affect aliasing, the same paragraph goes on to say that all pointers to structs must have the same representation, and presumably we wouldn't want *that* to affect aliasing.

Alright, I withdraw that comment; it's not formally required.  We may need to be more conservative about the aliasing of `void*` l-values on practical grounds, but we can let that be informed by results on the ground.

dybv-sc commented on clang/include/clang/Basic/CodeGenOptions.def:
```diff
@@ -216,6 +216,9 @@ ENUM_CODEGENOPT(StructReturnConvention, StructReturnConventionKind, 2, SRCK_Defa
 CODEGENOPT(StructPathTBAA    , 1, 0) ///< Whether or not to use struct-path TBAA.
+CODEGENOPT(UnionTBAA    , 1, 0) ///< Whether or not to use struct-path TBAA on unions.
+CODEGENOPT(PointerTBAA    , 1, 0) ///< Whether or not to generate TBAA on pointers.
+CODEGENOPT(ArrayTBAA    , 1, 0) ///< Whether or not to generate TBAA on arrays.
```
fixed

dybv-sc commented:
@rjmccall , I updated part of commit that handles pointers. Added comment with C/C++ standard references to explain my decisions. Could you please review it again?

nikic commented:
(As usual, please make any LLVM changes separately from Clang changes, especially if they affect IR design.)

Commit:
[IR][TBAA] Allow multiple fileds with same offset in TBAA struct-path

Support for multiple fields to have same offset in TBAA struct-path
metadata nodes. Primary goal is to support union-like structures
to participate in TBAA struct-path resolution.

Commit:
[Clang][CodeGen] Add TBAA metadata on pointer, union and array types.

    Options to disable new behaviour:

    -Xclang -no-pointer-tbaa
    -Xclang -no-array-tbaa

    Following option enables unions to participate in struct-path TBAA:

    -Xclang -union-tbaa

dybv-sc commented:
> (As usual, please make any LLVM changes separately from Clang changes, especially if they affect IR design.)

I splitted commit into 2 parts for llvm and clang. Should I make separate PRs for them as well?

dybv-sc commented:
I made separate PR for llvm's part: https://github.com/llvm/llvm-project/pull/76356
It should me merged first, because clang's part depends on it.

rjmccall commented on clang/lib/CodeGen/CodeGenTBAA.cpp:
```diff
@@ -184,13 +205,59 @@ llvm::MDNode *CodeGenTBAA::getTypeInfoHelper(const Type *Ty) {
-    return createScalarTypeNode("any pointer", getChar(), Size);
+  //
+  // In C11 for two pointer type to alias it is required for them to be
+  // compatible [section 6.5 p7].
```
This comment is pretty good, but it'd be better if we had a good comment on the function itself, and then this could be specifically in the context of that.  Please add something like this as a doc comment on the entire function:

```
/// Return an LLVM TBAA metadata node appropriate for an access through
/// an l-value of the given type.  Type-based alias analysis takes advantage
/// of the following rules from the language standards:
///
/// C 6.5p7:
///   An object shall have its stored value accessed only by an lvalue
///   expression that has one of the following types:
///   - a type compatible with the effective type of the object,
///   - a qualified version of a type compatible with the effective
///     type of the object,
///   - a type that is the signed or unsigned type corresponding
///     to the effective type of the object,
///   - a type that is the signed or unsigned type corresponding
///     to a qualified version of the effective type of the object,
///   - an aggregate or union type that includes one of the
///     aforementioned types among its members (including,
///     recursively, a member of a subaggregate or contained union), or
///   - a character type.
///
/// C++ [basic.lval]p11:
///   If a program attempts to access the stored value of an object
///   through a glvalue whose type is not similar to one of the following
///   types the behavior is undefined:
///   - the dynamic type of the object,
///   - a type that is the signed or unsigned type corresponding
///     to the dynamic type of the object, or
///   - a char, unsigned char, or std::byte type.
///
/// The C and C++ rules about effective/dynamic type are broadly similar
/// and permit memory to be reused with a different type.  C does not have
/// an explicit operation to change the effective type of memory; any store
/// can do it.  While C++ arguably does have such an operation (the standard
/// global `operator new(void*, size_t)`), in practice it is important to
/// be just as permissive as C.  We therefore treat all stores as being able to
/// change the effective type of memory, regardless of language mode.  That is,
/// loads have both a precondition and a postcondition on the effective
/// type of the memory, but stores only have a postcondition.  This imposes
/// an inherent limitation that TBAA can only be used to reorder loads
/// before stores.  This is quite restrictive, but we don't have much of a
/// choice.  In practice, hoisting loads is the most important optimization
/// for alias analysis to enable anyway.
///
/// Therefore, given a load (and its precondition) and an earlier store
/// (and its postcondition), the question posed to TBAA is whether there
/// exists a type that is consistent with both accesses.  If there isn't,
/// it's fine to hoist the load because either the memory is non-overlapping
/// or the precondition on the load is wrong (which would be UB).
///
/// LLVM TBAA says that two accesses with TBAA metadata nodes may alias if:
/// - the metadata nodes are the same,
/// - one of the metadata nodes is a base of the other (this can be
///   recursive, but it has to be the original node that's a base,
///   not just that the nodes have a common base), or
/// - one of the metadata nodes is a `tbaa.struct` node (the access
///   necessarily being a `memcpy`) with a subobject node that would
///   be allowed to alias with the other.
///
/// Our job here is to produce metadata nodes that will never say that
/// an alias is not allowed when there exists a type that would be consistent
/// with the types of the accesses from which the nodes were produced.
///
/// The last clause in both language rules permits character types to
/// alias objects of any type.  We handle this by converting all character
/// types (as well as `std::byte` and types with the `mayalias` attribute)
/// to a single metadata node (the `char` node), then making sure that
/// that node is a base of every other metadata node we generate.
/// We can always just conservatively use this node if we aren't otherwise
/// sure how to implement the language rules for a type.
///
/// Read literally, the C rule for aggregates permits an aggregate l-value
/// (e.g. of type `struct { int x; }`) to be used to access an object that
/// is not part of an aggregate object of that type (e.g. a local variable
/// of type `int`).  That case is perhaps sensical, but it would also permit
/// e.g. an l-value of type `struct { int x; float f; }` to be used to
/// access an object of type `float`, which is nonsense.  We interpret this
/// clause as just intending to permit objects to be accessed through an
/// l-value that properly references a containing object.
///
/// C++ does not have an explicit rule for aggregates because in C++
/// a non-member access to an aggregate l-value is always a call to a
/// constructor or assignment operator, which then accesses all the
/// subobjects.  In general, however, our interpretation of member
/// accesses is that they are also an access to the containing object
/// and therefore require such an object to exist at that address;
/// this permits us to just use the C rule for the accesses done by
/// trivial copy/move constructors/operators.
///
/// Both C and C++ permit some qualification differences.  In C, however,
/// qualification can only differ at the outermost level, whereas C++
/// allows qualification to differ in nested positions through the
/// similar-types rule.  This means that e.g. an l-value of type
/// `const float *` is not permitted to access an object of type
/// `float *` in C, but it is in C++.  We use the C++ rule
/// unconditionally; the C rule is needlessly strict and frequently
/// violated in practice by code that we don't want to say is wrong.
/// We implement this by just discarding type qualifiers within pointer-like
/// types when deriving TBAA nodes; basically, we produce the TBAA node
/// for the type that is unqualified at all the recursive positions
/// considered by the C++ similar type rule.  The implementation 
/// doesn't actually construct this recursively-qualified type as a
/// `QualType`; it just ignores qualifiers when recursing into types.
///
/// The similar-type rule only really applies to the standard CVR
/// qualifiers, which never affect representations.  Qualifiers such as
/// address spaces that may involve a representation difference would
/// be totally appropriate to distinguish for TBAA purposes.  However,
/// the current implementation just discards all qualifiers.
///
/// We handle the signed/unsigned clause by just making unsigned types
/// use the the metadata node for the signed variant of the type.  In the
/// language rules, this only applies at the outermost level, and e.g. an
/// l-value of type `signed int *` is not permitted to alias an object of
/// type `unsigned int *`.  We choose not to distinguish those types when
/// pointer-type TBAA is enabled, however.
///
/// After discarding qualifiers and signedness differences as above,
/// the language rules come down to whether the types are compatible
/// (in C) or identical (in C++).  Even in C, most types are compatible
/// only with themselves.  The exceptions will be considered in the cases
/// below.
```

and then this comment can just be something like this:

```
// When PointerTBAA is disabled, all pointers and references use the same
// "any pointer" TBAA node. Otherwise, we generate a type-specific TBAA
// node and use the "any pointer" node as its base for compatibility between
// TUs with different settings.  To implement the C++ similar-type rules
// (which we also adopt in C), we need to ignore qualifiers on the
// pointee type, and that has to be done recursively if the pointee type
// is itself a pointer-like type.
//
// Currently we ignore the differences between pointer-like types and just
// and use this tag for the type: `p<pointer depth> <inner type tag>`.
// This means we give e.g. `char **` and `char A::**` the same TBAA tag.
```

dybv-sc replied:
Great idea! Thanks for help with it. Added those comments.

rjmccall commented on clang/lib/CodeGen/CodeGenTBAA.cpp:
```diff
@@ -105,13 +105,34 @@ static bool isValidBaseType(QualType QTy) {
 }
 
+// Give unique tag for compatible types.
+std::string CodeGenTBAA::getPointeeName(const Type *Ty) {
```
Since this is always being used to build a larger string, could you make this append to a `raw_ostream`?

I guess `Ty` *is* guaranteed to be a canonical type here, because that's apparently a precondition of `getTypeInfoHelper`.  Could you assert that?

dybv-sc replied:
Changed function to append to raw_ostream. 
`Ty` here is coming from `getPointeeType()` so it may be not canonical, right?

fhahn commented:
Thanks for working on this! I think it would be good to split this up into multiple distinct parts for the different improvements.

I put up a patch to support distinct metadata for distinct pointers a while ago, which I just moved to GH: #76261.

The discussion on Phabricator has some interesting points and one of the concerns was that it is very difficult already to detect type violations in source code, which makes adopting more powerful TBAA features quite difficult. There are some tooling improvements we can make here, including a sanitizer for types, which would be good to get rolling again, see https://discourse.llvm.org/t/reviving-typesanitizer-a-sanitizer-to-catch-type-based-aliasing-violations/66092

TBAA union support and potential issues have been discussed in detail a number of years ago on the old mailing list (llvm-dev), did you have a look at those threads?

dybv-sc commented:
> Thanks for working on this! I think it would be good to split this up into multiple distinct parts for the different improvements.
> 
> I put up a patch to support distinct metadata for distinct pointers a while ago, which I just moved to GH: #76261.
> 
> The discussion on Phabricator has some interesting points and one of the concerns was that it is very difficult already to detect type violations in source code, which makes adopting more powerful TBAA features quite difficult. There are some tooling improvements we can make here, including a sanitizer for types, which would be good to get rolling again, see https://discourse.llvm.org/t/reviving-typesanitizer-a-sanitizer-to-catch-type-based-aliasing-violations/66092
> 
> TBAA union support and potential issues have been discussed in detail a number of years ago on the old mailing list (llvm-dev), did you have a look at those threads?

Hi,

Thanks for sharing those materials. Actually I didn't look into any discussions considering standard violations, so I would be glad if you share those threads.

Those standard violations worry me too, so as @rjmccall suggested, I put union TBAA under the option that is disabled by default. I am also considering to put pointer TBAA under option too, if it raises enough concerns too.

dybv-sc commented on clang/test/CXX/drs/dr158.cpp:
```diff
@@ -1,7 +1,7 @@
+// RUN: %clang_cc1 -triple x86_64-linux -std=c++14 %s -O3 -pedantic-errors -emit-llvm -o - | FileCheck %s
+// RUN: %clang_cc1 -triple x86_64-linux -std=c++1z %s -O3 -pedantic-errors -emit-llvm -o - | FileCheck %s
 
 // dr158: yes
```
As I see it is already mentioned:
https://github.com/llvm/llvm-project/blob/114e6d7ba02f090117f2cb1ffeb9027cf80f335b/clang/test/CXX/drs/dr1xx.cpp#L792C1-L792C25

Endilll replied:
I'm sorry I missed that you're not writing a new test.

thesamesam commented:
> I put union TBAA under the option that is disabled by default. I am also considering to put pointer TBAA under option too, if it raises enough concerns too.

Let's please file a bug once this is merged so we remember to come back to that.
